components/script/dom/node.rs

```diff
@@ -1230,66 +1271,76 @@ impl Iterator for PrecedingNodeIterator {
         } else if let Some(previous_sibling) = current.GetPreviousSibling() {
             if self.root == previous_sibling {
                 None
-            } else if let Some(last_child) = previous_sibling.descending_last_children().last() {
+            } else if let Some(last_child) = previous_sibling.descending_last_children::<Node>().last() {
                 Some(last_child)
             } else {
                 Some(previous_sibling)
             }
         } else {
             current.GetParentNode()
         };
-        self.current.clone()
+
+        //TODO: dominotree - not sure how to handle failures here - is returning None okay if it can't cast to T?
+        if let Some(current_node) = self.current.take() {
+            Root::downcast::<T>(current_node)
+        } else {
+            None
+        }
     }
 }
 
-pub struct LastChildIterator {
+pub struct LastChildIterator<T: DerivedFrom<Node>> {
     current: Option<Root<Node>>,
+    phantom: PhantomData<T>,
 }
 
-impl Iterator for LastChildIterator {
-    type Item = Root<Node>;
+impl<T> Iterator for LastChildIterator<T> where T: DerivedFrom<Node> {
+    type Item = Root<T>;
 
-    fn next(&mut self) -> Option<Root<Node>> {
+    fn next(&mut self) -> Option<Self::Item> {
         let current = match self.current.take() {
             None => return None,
             Some(current) => current,
         };
         self.current = current.GetLastChild();
-        Some(current)
+        Root::downcast::<T>(current.clone())
     }
 }
 
-pub struct AncestorIterator {
+pub struct AncestorIterator<T: DerivedFrom<Node>> {
     current: Option<Root<Node>>,
+    phantom: PhantomData<T>,
 }
 
-impl Iterator for AncestorIterator {
-    type Item = Root<Node>;
+impl<T> Iterator for AncestorIterator<T> where T: DerivedFrom<Node> {
+    type Item = Root<T>;
 
-    fn next(&mut self) -> Option<Root<Node>> {
+    fn next(&mut self) -> Option<Self::Item> {
         let current = match self.current.take() {
             None => return None,
             Some(current) => current,
         };
         self.current = current.GetParentNode();
-        Some(current)
+        Root::downcast::<T>(current.clone())
```

Comment:
That's wrong, instead of returning `None`, it should try to return the ancestor of `self.current`, etc.